tpm2/structures.go

```diff
@@ -668,111 +689,345 @@ type TPMSACTData struct {
 }
 
 // TPMLCCA represents a TPML_CCA.
 // See definition in Part 2: Structures, section 10.9.2.
 type TPMLCCA struct {
+	marshalByReflection
 	CommandAttributes []TPMACC `gotpm:"list"`
 }
 
 // TPMLAlg represents a TPMLALG.
 // See definition in Part 2: Structures, section 10.9.3.
 type TPMLAlg struct {
+	marshalByReflection
 	Algorithms []TPMAlgID `gotpm:"list"`
 }
 
 // TPMLHandle represents a TPML_HANDLE.
 // See definition in Part 2: Structures, section 10.9.4.
 type TPMLHandle struct {
+	marshalByReflection
 	Handle []TPMHandle `gotpm:"list"`
 }
 
 // TPMLDigest represents a TPML_DIGEST.
 // See definition in Part 2: Structures, section 10.9.5.
 type TPMLDigest struct {
+	marshalByReflection
 	// a list of digests
 	Digests []TPM2BDigest `gotpm:"list"`
 }
 
 // TPMLDigestValues represents a TPML_DIGEST_VALUES.
 // See definition in Part 2: Structures, section 10.9.6.
 type TPMLDigestValues struct {
+	marshalByReflection
 	// a list of tagged digests
 	Digests []TPMTHA `gotpm:"list"`
 }
 
 // TPMLPCRSelection represents a TPML_PCR_SELECTION.
 // See definition in Part 2: Structures, section 10.9.7.
 type TPMLPCRSelection struct {
+	marshalByReflection
 	PCRSelections []TPMSPCRSelection `gotpm:"list"`
 }
 
 // TPMLAlgProperty represents a TPML_ALG_PROPERTY.
 // See definition in Part 2: Structures, section 10.9.8.
 type TPMLAlgProperty struct {
+	marshalByReflection
 	AlgProperties []TPMSAlgProperty `gotpm:"list"`
 }
 
 // TPMLTaggedTPMProperty represents a TPML_TAGGED_TPM_PROPERTY.
 // See definition in Part 2: Structures, section 10.9.9.
 type TPMLTaggedTPMProperty struct {
+	marshalByReflection
 	TPMProperty []TPMSTaggedProperty `gotpm:"list"`
 }
 
 // TPMLTaggedPCRProperty represents a TPML_TAGGED_PCR_PROPERTY.
 // See definition in Part 2: Structures, section 10.9.10.
 type TPMLTaggedPCRProperty struct {
+	marshalByReflection
 	PCRProperty []TPMSTaggedPCRSelect `gotpm:"list"`
 }
 
 // TPMLECCCurve represents a TPML_ECC_CURVE.
 // See definition in Part 2: Structures, section 10.9.11.
 type TPMLECCCurve struct {
+	marshalByReflection
 	ECCCurves []TPMECCCurve `gotpm:"list"`
 }
 
 // TPMLTaggedPolicy represents a TPML_TAGGED_POLICY.
 // See definition in Part 2: Structures, section 10.9.12.
 type TPMLTaggedPolicy struct {
+	marshalByReflection
 	Policies []TPMSTaggedPolicy `gotpm:"list"`
 }
 
 // TPMLACTData represents a TPML_ACT_DATA.
 // See definition in Part 2: Structures, section 10.9.13.
 type TPMLACTData struct {
+	marshalByReflection
 	ACTData []TPMSACTData `gotpm:"list"`
 }
 
+type tpmuCapabilities struct {
+	selector TPMCap
+	contents Marshallable
+}
+
+type capabilitiesContents interface {
+	Marshallable
+	*TPMLAlgProperty | *TPMLHandle | *TPMLCCA | *TPMLCC | *TPMLPCRSelection | *TPMLTaggedTPMProperty |
+		*TPMLTaggedPCRProperty | *TPMLECCCurve | *TPMLTaggedPolicy | *TPMLACTData
+}
+
+// create implements the UnmarshallableWithHint interface.
+func (u *tpmuCapabilities) create(hint int64) (reflect.Value, error) {
```

Comment:
I'm not 100% happy with this pattern for TPMUs yet. There are few enough TPMU structures that I'm not overly worried about this, but it would be nice to have a helper type or function to make create and get more brief.

Part of the problem is that `reflect.ValueOf` needs to be called on a variable whose type is the actual one we want, and not the interface (e.g., `Marshallable`) so something somewhere needs to instantiate contents on a per-type basis.

It might be possible to create a helper that holds onto a map of selector values to prototype contents and looks it up in the map, but then we have to worry about when that map is allocated (since a zero-valued TPMU needs to be able to return the zero-valued contents of the requested union member without initialization)

Reply:
Haven't thought of a better pattern, and since there are only a few TPMUs, the cost of changing this later is pretty low.